### PR TITLE
implements "include required"

### PR DESCRIPTION
**Pull request by Johnlon**

### CODE REVIEW

**Comment by Johnlon** [comment]
This pull requests adds the "required" include syntax discussed in https://github.com/typesafehub/config/issues/121

File: README.md [hunk 1]
```diff
@@ -840,6 +840,7 @@ format.
   * config-annotation https://github.com/wacai/config-annotation
   * PureConfig https://github.com/melrief/pureconfig
   * Simple Scala Config https://github.com/ElderResearch/ssc
+  * konfig https://github.com/vpon/konfig
```
Comment (havocp) [inline README.md]: this is unrelated I guess, maybe need to rebase on master ?

File: config/src/main/java/com/typesafe/config/impl/Tokenizer.java [hunk 2]
```diff
@@ -481,7 +481,7 @@ private Token pullQuotedString() throws ProblemException {
             // the open quote has already been consumed
             StringBuilder sb = new StringBuilder();
 
-            // We need a second string builder to keep track of escape characters.
+            // We need a sec ond string builder to keep track of escape characters.
             // We want to return them exactly as they appeared in the original text,
```
Comment (havocp) [inline config/src/main/java/com/typesafe/config/impl/Tokenizer.java]: typo

File: config/src/main/java/com/typesafe/config/impl/Tokenizer.java [hunk 3]
```diff
@@ -299,7 +299,7 @@ private Token pullComment(int firstChar) {
         // chars JSON allows to be part of a number
         static final String numberChars = "0123456789eE+-.";
         // chars that stop an unquoted string
-        static final String notInUnquotedText = "$\"{}[]:=,+#`^?!@*&\\";
+        static final String notInUnquotedText = "$\"{}()[]:=,+#`^?!@*&\\";
```
Comment (havocp) [inline config/src/main/java/com/typesafe/config/impl/Tokenizer.java]: I don't remember if something like `foo = (hello)` would parse as `foo = "(hello)"` before, but if so we need it to continue to do so. We should be sure there's a test case for parsing or failing to parse parens in unquoted strings.

File: config/src/test/scala/com/typesafe/config/impl/ConfParserTest.scala [hunk 4]
```diff
@@ -780,6 +781,45 @@ class ConfParserTest extends TestUtils {
     }
 
     @Test
+    def includeRequiredMissing() {
```
Comment (havocp) [inline config/src/test/scala/com/typesafe/config/impl/ConfParserTest.scala]: should we test a third level of nesting (that required() does not affect the requiredness of includes inside the required file - it should always affect only the immediate include)

File: config/src/test/scala/com/typesafe/config/impl/TestUtils.scala [hunk 5]
```diff
@@ -721,8 +721,10 @@ abstract trait TestUtils {
 
     val resourceDir = {
         val f = new File("config/src/test/resources")
-        if (!f.exists())
-            throw new Exception("Tests must be run from the root project directory containing " + f.getPath())
+        if (!f.exists()) {
+            val here = new File(".").getAbsolutePath
+            throw new Exception(s"Tests must be run from the root project directory containing ${f.getPath()}, however the current director is $here")
+        }
```
Comment (havocp) [inline config/src/test/scala/com/typesafe/config/impl/TestUtils.scala]: director missing a y

**Comment by havocp** [comment]
thanks! looks very close to done, I found some nitpicks.

I think propagating the allowMissing flag to includes within the required file might be an easy bug to introduce accidentally so we might want to test that third level of nesting.

Also we should be confident that making parens into a token hasn't created any incompatible changes. I'm not sure the test coverage includes this case already, it might though. I'm sure there are tests related to other tokens and chars that show the idea if not. I think somewhere there are tests with a big list of valid and invalid strings or something and adding stuff like "()" to those lists would get us coverage.

it might be good to add a required() to the load/save roundtrip tests (ConfigDocumentFactory) to be sure we can round trip an include(required()) statement.

File: config/src/main/java/com/typesafe/config/impl/Tokenizer.java (hunk 3 above)
Comment (Johnlon) [inline config/src/main/java/com/typesafe/config/impl/Tokenizer.java]: What should "foo = (hello)" mean?

Comment (havocp) [inline config/src/main/java/com/typesafe/config/impl/Tokenizer.java]: I don't know offhand if parens were allowed in unquoted strings before, but if so they should continue to be allowed.

Comment (havocp) [inline config/src/main/java/com/typesafe/config/impl/Tokenizer.java]: I think they were allowed in unquoted strings before. Could add tests here for example https://github.com/typesafehub/config/blob/b076de21ceaa41789c80870609cb7435e09f2356/config/src/test/scala/com/typesafe/config/impl/TestUtils.scala#L423 and also in the tokenizer test if they are now parsed as tokens.

**Comment by Johnlon** [comment]
the tests that failed in Travis are coded not to run on Windows.
I removed the exclusion and fixed the tests.

File: config/src/main/java/com/typesafe/config/impl/Tokenizer.java (hunk 3 above)
Comment (Johnlon) [inline config/src/main/java/com/typesafe/config/impl/Tokenizer.java]: _(hello)_ previously parsed as _tokenUnquoted("(footrue)")_

Docs say .. 

_A sequence of characters outside of a quoted string is a string value if:
it does not contain "forbidden characters": '$', '"', '{', '}', '[', ']', ':', '=', ',', '+', '#', '`', '^', '?', '!', '@', '*', '&', '\' (backslash), or whitespace._

.. so of course no mention of () but then the behaviour of () outside "" isn't explicit

Comment (Johnlon) [inline config/src/main/java/com/typesafe/config/impl/Tokenizer.java]: Reviewing the structure of the distribution of logic across the tokeniser and the doc parser etc its not clear how () may be both as tokens at the same time have _(hello)_ be a single unquoted string.

I will withdraw the implementation entirely.

Pity really because the pull request I made 2 years ago worked fine but that was before the tokeniser was introduced. :(

**Comment by Johnlon** [comment]
The failing test is the one that highlights that there is a conflict between () tokenisation and (hello) as an unquotedString.

Don't know how to resolve this so perhaps best just to reject and close this pull request.
